### ISSUE TITLE
[fabricjs] Update type definitions agains Fabric 4.1.0

### ISSUE THREAD

**Issue by DanielPower** [DefinitelyTyped/DefinitelyTyped#47705]
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @oklemencic @joewashear007 @mrand01 @NotWoods @bmartinson @RogerioTeixeira @BradleyHill @bmkrol823 @glenngartner @codertx @mike667 @nataliemarleny 

Fabric 4.0.0 is a new major release which includes a significant new feature, custom control points. There are currently no type definitions for this.

**Comment by dkaczor97** [comment]
Any news? It's been a year since the release and I still cannot see custom controls support.

**Comment by DanielPower** [comment]
@types/fabric has been bumped to version 4.2.2. I haven't had a chance to test yet if that includes type definitions for new features like control points. @dkaczor97 have you tried 4.2.2?

**Comment by dkaczor97** [comment]
@DanielPower Yes, I've checked it. Still no type definition for custom controls.

**Comment by asturur** [comment]
Hi Everyone, as of now i maintain fabricjs.
In the issues on the main repo, often people ask about type definitions, someone send people here, and someone else just complains that those definitions aren't perfect.

I m not that much near to the typescript community and habits, and i would like to understand the following things.

Is it preferable for TS users to have a type definition file in the repository with the library or in a generic repository like this one?

If is preferable in the main repository, is it ok to move the current file there and be sure is kept updated with some CI utils commit by commit?

**Comment by doouding** [comment]
> Hi Everyone, as of now i maintain fabricjs.
> In the issues on the main repo, often people ask about type definitions, someone send people here, and someone else just complains that those definitions aren't perfect (but also they don't do anything to improve it 🤷 )
> 
> I m not that much near to the typescript community and habits, and i would like to understand the following things.
> 
> Is it preferable for TS users to have a type definition file in the repository with the library or in a generic repository like this one?
> 
> If is preferable in the main repository, is it ok to move the current file there and be sure is kept updated with some CI utils commit by commit?

I think have a type definition file in the repository is better. Editors like vscode can load the type definition file automatically and makes code completion more accurate (even the code is written in javascript). And of course, it may takes some effort to maintain the type and keep update with the latest api.

**Comment by DanielPower** [comment]
I think it makes sense to ship the type declarations in the `fabric` repo if they're going to be maintained by the fabric maintainers. Which doesn't seem like it will be the case unless they can be generated automatically, since @asturur isn't interested in writing them - which is fine, I appreciate the work you put into fabric, and wouldn't expect you to take on the extra workload.

@asturur you fairly heavily use JSDoc in fabric. I'm not sure how good the resulting type definitions would be, but Typescript does support auto-generating declaration files from JSDoc comments. So that's one thing you could try if you wanted to provide official typescript declarations, but don't want to have to manually maintain them. You could set it to auto-generate via CI.

https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html
https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

It may not be perfect at first, but by using JSDoc generated type declarations, any community efforts to improve Fabric's type declarations would also be directed at improving your existing documentation / JSDoc comments. So I see that as a potential win.

Again, as a disclaimer, I have no experience with projects that use this. So I'd love it if someone who's actually used JSDoc generated type declarations could chime in about any downsides or nuances.

**Comment by asturur** [comment]
So is not that i m not willing to do so, i simply do not have time to start a blank file from scratch. Is a daunting task that i will postpone forever.
Different would be if i had a file to tweak and a CI utility that would complain if something does not match.

**Comment by asturur** [comment]
Someone was looking into generating from JSDOC but also got issues because fabric is declared inside inline function executed at runtime and seems that this disturb the parser
